xdsl/interpreters/onnx.py

```diff
@@ -66,3 +68,221 @@ def run_relu(self, interpreter: Interpreter, op: onnx.Relu, args: tuple[Any, ...
+        operand_data = np.array(operand.data)
+        return ShapedArray(list(operand_data), new_shape)
+
+    @impl(onnx.Gemm)
+    def run_gemm(self, interpreter: Interpreter, op: onnx.Gemm, args: tuple[Any, ...]):
+        a, b, c = args[0], args[1], args[2]
+
+        alpha = op.alpha.value.data if op.alpha is not None else 1.0
+        beta = op.beta.value.data if op.beta is not None else 1.0
+
+        assert isinstance(a, ShapedArray)
+        assert isinstance(b, ShapedArray)
+        assert isinstance(c, ShapedArray)
+
+        a = cast(ShapedArray[float], a)
+        b = cast(ShapedArray[float], b)
+        c = cast(ShapedArray[float], c)
+
+        a = np.array(a.data)
+        b = np.array(b.data)
+        c = np.array(c.data)
+
+        if op.trans_a is not None and op.trans_a.value.data == 1:
+            a = np.transpose(a)
+
+        if op.trans_b is not None and op.trans_b.value.data == 1:
+            b = np.transpose(b)
+
+        result = alpha * np.dot(a, b) + beta * c
+
+        result_type = op.res_tensor.type
+        result_shape = list(result_type.get_shape())
+        assert isinstance(result_type, TensorType)
+
+        return ShapedArray(list(result), result_shape)
+
+    @impl(onnx.Conv)
+    def run_conv(self, interpreter: Interpreter, op: onnx.Conv, args: tuple[Any, ...]):
+
+        # initialise the attributes used
+        auto_pad = op.auto_pad.data
+        strides: list[int] = [value.value.data for value in op.strides]
+        matrix, kernel, bias = args[0], args[1], args[2]
+        pads: list[int] = [value.value.data for value in op.pads]
+
+        matrix = cast(ShapedArray[float], matrix)
+        kernel = cast(ShapedArray[float], kernel)
+        bias = cast(ShapedArray[float], bias)
+
+        matrix = np.array(matrix.data).reshape(matrix.shape)
+        kernel = np.array(kernel.data).reshape(kernel.shape)
+
+        if auto_pad != "NOTSET":
+            if auto_pad == "SAME_UPPER" or auto_pad == "SAME_LOWER":
+                out_height = np.ceil(matrix.shape[2] / strides[0])
+                out_width = np.ceil(matrix.shape[3] / strides[1])
+
+                pad_along_height = max(
+                    (out_height - 1) * strides[0] + kernel.shape[2] - matrix.shape[2], 0
+                )
+                pad_along_width = max(
+                    (out_width - 1) * strides[1] + kernel.shape[3] - matrix.shape[3], 0
+                )
+
+                if auto_pad == "SAME_UPPER":
+                    pad_top = int(pad_along_height // 2)
+                    pad_bottom = int(pad_along_height - pad_top)
+                    pad_left = int(pad_along_width // 2)
+                    pad_right = int(pad_along_width - pad_left)
+                else:
+                    pad_bottom = int(pad_along_height // 2)
+                    pad_top = int(pad_along_height - pad_bottom)
+                    pad_right = int(pad_along_width // 2)
+                    pad_left = int(pad_along_width - pad_right)
```

Comment:
why the ints? I think they're guaranteed to be ints already

Reply:
for the function `test_onnx_conv_same_lower_strides` without the ints the pad width is not of integer type

Reply:
oh then I'd suggest changing the lines for out_height and out_width to add the `int` initialisation, so the downstream calculations are all strictly on integers

Reply:
they do have the int stuff

Reply:
I don't see it:
```

                out_height = np.ceil(matrix.shape[2] / strides[0])
                out_width = np.ceil(matrix.shape[3] / strides[1])
```

Reply:
oh lol, i was looking at the other one, its added now